goalc/compiler/compilation/CompilerControl.cpp

```diff
@@ -616,3 +628,185 @@ Val* Compiler::compile_print_debug_compiler_stats(const goos::Object& form,
 
   return get_none();
 }
+
+Val* Compiler::compile_gen_docs(const goos::Object& form, const goos::Object& rest, Env*) {
+  auto args = get_va(form, rest);
+  va_check(form, args, {goos::ObjectType::STRING}, {});
+
+  const auto& doc_path = fs::path(args.unnamed.at(0).as_string()->data);
+  lg::info("Saving docs to: {}", doc_path.string());
+
+  const auto symbols = m_symbol_info.get_all_symbols();
+
+  std::unordered_map<std::string, Docs::SymbolDocumentation> all_symbols = {};
+  std::unordered_map<std::string, Docs::FileDocumentation> file_docs = {};
+
+  lg::info("Processing {} symbols...", symbols.size());
+  int count = 0;
+  for (const auto& sym_info : symbols) {
+    count++;
+    if (count % 100 == 0 || count == symbols.size()) {
+      lg::info("Processing [{}/{}] symbols...", count, symbols.size());
+    }
+    std::optional<Docs::DefinitionLocation> def_loc;
+    const auto& goos_info = m_goos.reader.db.get_short_info_for(sym_info.src_form());
+    if (goos_info) {
+      Docs::DefinitionLocation new_def_loc;
+      new_def_loc.filename = file_util::convert_to_unix_path_separators(file_util::split_path_at(
+          goos_info->filename, {"goal_src", version_to_game_name(m_version)}));
+      new_def_loc.line_idx = goos_info->line_idx_to_display;
+      new_def_loc.char_idx = goos_info->pos_in_line;
+      def_loc = new_def_loc;
+    }
+
+    Docs::SymbolDocumentation sym_doc;
+    sym_doc.name = sym_info.name();
+    sym_doc.description = sym_info.meta().docstring;
+    sym_doc.kind = sym_info.kind();
+    sym_doc.def_location = def_loc;
+
+    if (all_symbols.count(sym_info.name()) > 1) {
+      lg::error("A symbol was defined twice, how did this happen? {}", sym_info.name());
+    } else {
+      all_symbols.emplace(sym_info.name(), sym_doc);
+    }
+
+    Docs::FileDocumentation file_doc;
+    std::string file_doc_key;
+    if (!goos_info) {
+      file_doc_key = "unknown";
+    } else {
+      file_doc_key = file_util::convert_to_unix_path_separators(
+          file_util::split_path_at(goos_info->filename, {"goal_src"}));
+    }
+
+    if (file_docs.count(file_doc_key) != 0) {
+      file_doc = file_docs.at(file_doc_key);
+    } else {
+      file_doc = Docs::FileDocumentation();
+    }
+
+    // TODO - states / enums / built-ins
+    if (sym_info.kind() == SymbolInfo::Kind::GLOBAL_VAR ||
+        sym_info.kind() == SymbolInfo::Kind::CONSTANT) {
+      Docs::VariableDocumentation var;
+      var.name = sym_info.name();
+      var.description = sym_info.meta().docstring;
+      if (sym_info.kind() == SymbolInfo::Kind::CONSTANT) {
+        var.type = object_type_to_type_name(m_global_constant_types.at(var.name).type);
```

Comment:
You can't really know the type of a constant - they are just replaced as text directly so it can be context-dependent. For example, the type of `foo` here can be both string or int.
```
g > (defglobalconstant foo bar)
g > (let ((bar "a")) (print-type foo))
[TYPE] string igpr-1
g > (let ((bar 1)) (print-type foo))
[TYPE] int igpr-1
```

that said, I don't think we abuse constants like this in any existing code, so maybe it's ok if these are wrong.